### PR TITLE
extending Cypress logging when the regression is running 

### PR DESCRIPTION
**Pull request by ScriptingShrimp**
### Describe the change

When Cypress regression fails, its a bit tedious to understand why. Usually first steps to understand the issue are to reproduce the issue locally and catch the Kiali pod logs (eventually logs form other pods). 

Having this logs available for each test run (now per feature file) can save some time in this process and give early insights  without whole routine of reproducing locally.

Here is log output example for `wizard_istio_config.feature.json`
```json
{"message":"404 Not Found","nodeName":"ci-worker","namespace":"bookinfo","podName":"kiali-traffic-generator-5chns","containerName":"kiali-traffic-generator"}
{"message":"2025-06-24T17:31:13Z DBG Validating token session","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13Z DBG Scraping istiod for debug info","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13Z TRC Getting proxy status from istiod in cluster [Kubernetes] for revision [default]","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13Z TRC Getting services from istiod in cluster [Kubernetes] for revision [default]","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13Z INF AUDIT: Object: {\"kind\":\"Gateway\",\"apiVersion\":\"networking.istio.io/v1\",\"metadata\":{\"name\":\"mygateway\",\"namespace\":\"bookinfo\",\"labels\":{},\"annotations\":{}},\"spec\":{\"selector\":{\"istio\":\"ingressgateway\"},\"servers\":[{\"port\":{\"number\":8080,\"name\":\"foobar\",\"protocol\":\"HTTP\"},\"hosts\":[\"website.com\"],\"tls\":{}}]}} group=istioConfig gvk=\"networking.istio.io/v1, Kind=Gateway\" namespace=bookinfo operation=CREATE request-id=d1de1oct3kjc73fp09bg route=IstioConfigCreate route-pattern=/api/namespaces/{namespace}/istio/{group}/{version}/{kind} user=istio-system:citest","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13.199969Z\tinfo\tads\tPush debounce stable[42] 1 for config Gateway/bookinfo/mygateway: 100.627336ms since last change, 100.627296ms since last push, full=true","nodeName":"ci-worker","namespace":"istio-system","podName":"istiod-54766b954f-4b6xg","containerName":"discovery"}
{"message":"2025-06-24T17:31:13.200149Z\tinfo\tads\tXDS: Pushing Services:33 ConnectedEndpoints:42 Version:2025-06-24T17:31:13Z/13","nodeName":"ci-worker","namespace":"istio-system","podName":"istiod-54766b954f-4b6xg","containerName":"discovery"}
{"message":"2025-06-24T17:31:13.200415Z\tinfo\tdelta\tLDS: PUSH for node:k8sapigateway-istio-74f6f44fd6-v25jw.bookinfo resources:0 removed:0 size:0B","nodeName":"ci-worker","namespace":"istio-system","podName":"istiod-54766b954f-4b6xg","containerName":"discovery"}
{"message":"2025-06-24T17:31:13.200603Z\tinfo\tdelta\tLDS: PUSH for node:istio-ingressgateway-86cdf57597-2hzbp.istio-system resources:1 removed:0 size:2.3kB","nodeName":"ci-worker","namespace":"istio-system","podName":"istiod-54766b954f-4b6xg","containerName":"discovery"}
{"message":"2025-06-24T17:31:13.200770Z\tinfo\tdelta\tRDS: PUSH for node:istio-ingressgateway-86cdf57597-2hzbp.istio-system resources:1 removed:0 size:2.4kB cached:0/0","nodeName":"ci-worker","namespace":"istio-system","podName":"istiod-54766b954f-4b6xg","containerName":"discovery"}
{"message":"2025-06-24T17:31:13Z DBG Validating token session","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13Z DBG Validating token session","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13Z DBG Validating token session","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13Z TRC Istio components config not set. Autodetecting components.","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13Z TRC GetGateways: Returning list from cache","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
{"message":"2025-06-24T17:31:13Z ERR Error fetching availability of the tracing service: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing: dial tcp 10.96.90.229:16685: connect: connection refused\"","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6d98bc6b49-lkxsj","containerName":"kiali"}
```

### Steps to test the PR

- clone this branch/PR
- `yarn cypress` and run some tests 

### Automation testing

extend GH actions to gather this log files 

### Issue reference

will create issue for this

### CODE REVIEW

**Comment by nrfox** [comment]
The Kiali pod logs are included on failure under artifacts --> debug info:
Screenshot from 2025-06-20 12-46-24

**Comment by ScriptingShrimp** [comment]
@nrfox please can you give me second thoughts?

**Comment by nrfox** [comment]
> @nrfox please can you give me second thoughts?

Log files are already being captured on failure in GH actions. What is the value add with this change?

**Comment by ScriptingShrimp** [comment]
Lately, Kiali application improved its logging. This builds on top of that.

> Log files are already being captured on failure in GH actions. What is the value add with this change?

@nrfox In the generated artifacts I can see logs from
```
kiali_logs.txt   
metallb_controller_current_logs.txt
metallb_speaker_current_logs.txt
``` 
besides that, there are additional outputs  
```
kiali_pods.yaml  
describe_nodes.txt 
describe_metallb_pods.txt  
```

All of this files are created as post action using `hack/ci-get-debug-info.sh` script. I see following pragmatic problems with that:

- you need to rerun the failed GH action with the `enable debug` option first to get those 
- the log files are not in unified format (stern outputs JSON for all scraped resources)
- with multiple log files (without timestamps), you loose context and its hard to find relevant events across multiple files 
- if pods gets restarted (crashed or reconciled with different NAME-ID) the logs are gone from that pod forever
- this approach will work currently for the upstream pipelines and with close to none changes in the private pipelines 

Although some issues described above could be fiexd, there is more.
benefits of this approach:

- when you are debugging Cypress locally (outside of GH actions), you don't need to have custom setup to tail the resources. Logs will get created automatically for each run
- logs are generated in separate files, unified format, with timestamps - chronologically, as tests were executed 
- each file is divided by `--scenario--` placeholder (this needs some improvements) which can help you identify events related to failed testcase
- upstream first, no skeleton in downstream closet 

future plans: 

- unified JSON format will enable us store and analyze logs - on a team call with @jshaughn , we concluded it will be interesting to observe the application under constant load
- trying to log "every" resource can uncover hidden errors (various platforms, enviroments)
- this approach is more flexible with scraping relevant resources during the test execution (i.e. tests doing changes on `sleep` , we can attach stern on `sleep` )

As Kiali application grows, it interacts with more resources via its *excellent* regression suite. Generally, this should help us to navigate better in test results from those extra resources. I also updated steps to reproduce, so please give it a try.

File: hack/stern/download-stern.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  # Get latest version tag from GitHub API
+  VERSION=$(curl -s "https://api.github.com/repos/$REPO/releases/latest" | $GREP_CMD -oP '"tag_name": "\K(.*)(?=")' | sed 's/^v//')
+
+  # Define the filename and URL
+  FILENAME="stern_${VERSION}_${OS}_${ARCH}.tar.gz"
+  URL="https://github.com/${REPO}/releases/download/v${VERSION}/${FILENAME}"
+
+  echo $URL
+  echo $FILENAME
```
Comment (ScriptingShrimp) [inline hack/stern/download-stern.sh]: remove

**Comment by nrfox** [comment]
Thanks for spelling out more of the requirements as I'm trying to understand what the needs are. It sounds like it'd be useful to have more context in the logs pulled from the environment and I definitely agree with that.

Some of the needs are already or can be addressed without additional dependencies:
> you need to rerun the failed GH action with the enable debug option first to get those

You don't need to re-run the action in debug mode to see the logs artifacts. See this failed run as an example: https://github.com/kiali/kiali/actions/runs/15825654603.

> the log files are not in unified format (stern outputs JSON for all scraped resources)

Yes they're just text but IMO text is easier to read than JSON anyways. If we're piping the logs into some analysis tool then we can convert them to JSON at that step.

> with multiple log files (without timestamps), you loose context and its hard to find relevant events across multiple files

This `kubectl logs --help` has a few options that meets a lot of these needs. e.g.
```
kubectl logs -l app=reviews -n bookinfo --all-containers --tail=-1 -f --max-log-requests 10 --timestamps=true --prefix=true
```
This will show all logs for pods matching the `app=review` label in `bookinfo` while adding timestamps and pod name prefixes. What does `stern` add that you can't already do with `kubectl logs`?

I think streaming some logs per feature file would be useful provided it doesn't slow down the execution of the tests or consume too many resources on the already limited CI environment.

**Comment by ScriptingShrimp** [comment]
I do agree your point of view. Namely JSON reading is not pleasant. For local workflows I'll provide better solution later. 

> What does stern add that you can't already do with kubectl logs?

- if pods gets restarted (crashed or reconciled with different NAME-ID) the logs are gone from that pod forever

Stern is just binary which do extend `kubectl logs`. Originally I was considering to just use kubectl, but `kubectl logs -f` will fail in scenario described above. Honestly, I don't want to reinvent wheel here - by catching all edge cases when kubectl might fail and sanitize that in bash. 

Only concern I have that we will hit github download pages with each regression run, but this could be solved later with the test container having this tool bundled in it.

**Comment by nrfox** [comment]
> I do agree your point of view. Namely JSON reading is not pleasant. For local workflows I'll provide better solution later.

It's not just local workflows. We don't pipe the logs into any kind of analysis tools now so there's no way to take advantage of the change in format when the CI pipeline fails.

> if pods gets restarted (crashed or reconciled with different NAME-ID) the logs are gone from that pod forever

I think you'd still get logs from pods that are crashing as long as they existed prior to running `kubectl logs`. What you don't get that `stern` might add is that when new pods are added that match the selector, they are not added to the log stream.

> Only concern I have that we will hit github download pages with each regression run, but this could be solved later with the test container having this tool bundled in it.

If we are adding this dep to get the feature of new pods being added to the log stream, then it needs to be treated like the other tools (istioctl, golang linter) and downloaded and installed outside of the test suite. I'm still not convinced we need the dep but we should improve the logging and having logging per feature file sounds useful too.

**Comment by ScriptingShrimp** [comment]
> It's not just local workflows. We don't pipe the logs into any kind of analysis tools now so there's no way to take advantage of the change in format when the CI pipeline fails.

Of course we don't pipe any logs, because we don't have them. There is no requirement to do that. It's more like devops thing to do that. Having JSON artifacts is first step to enable that.   

> you'd still get logs from pods that are crashing

Yes and no. Depends `kubectl logs --help`
```
  # Return snapshot of previous terminated ruby container logs from pod web-1
  kubectl logs -p -c ruby web-1
```
how many times the container was restarted. Once you don't tail on current resources and rely on `-p` flag, you will only get the logs from the n-1 container. This was something I was dealing with when debugging downstream pipeline. 

> stern might add is that when new pods are added that match the selector

Exactly. I encourage you to try that and see how it works. From the docs 
```
The query is a regular expression or a Kubernetes resource in the form <resource>/<name> so the pod name can easily be filtered and you don't need to specify the exact id (for instance omitting the deployment id)
```

>  treated like the other tools (istioctl, golang linter)

No. This is now test (dev) dependency only. To avoid any false positives caused by this tool, I can introduce global flag for the tests to enable or disable this feature. IMO enabled by default. We can disable it if we face any problems, until we find out what is wrong - than re-enable it again. Having meaningful traces from the regression is the goal.

**Comment by nrfox** [comment]
> Of course we don't pipe any logs, because we don't have them. There is no requirement to do that. It's more like devops thing to do that. Having JSON artifacts is first step to enable that.

We do have logs and they can help in debugging CI failures. If we switch everything to JSON it makes them harder to read as a human. You can always convert them to JSON from text if you want to do pipe them into another tool. Or we could even have two log stream, one JSON and one text but converting everything to JSON now would only make it harder to read the logs and to debug CI failure and not easier.

> how many times the container was restarted. Once you don't tail on current resources and rely on -p flag, you will only get the logs from the n-1 container. This was something I was dealing with when debugging downstream pipeline.

Yes this is the main feature that stern provides that `kubectl logs` does not.

> No. This is now test (dev) dependency only. To avoid any false positives caused by this tool, I can introduce global flag for the tests to enable or disable this feature.

The rest of the tools are also test/dev dependencies only but these are not installed _inside of_ the test suite. We don't want to add another flag. If we are adding another dev/test tool then it needs to be installed outside of the test suite. The log output(s) should also be included in the CI artifacts that you can download from CI failures. There's a script that gathers up the current info: https://github.com/kiali/kiali/blob/master/hack/ci-get-debug-info.sh and you can see where this is called in the CI workflow on failure: https://github.com/kiali/kiali/blob/3e2e9648e4ccd248bed20a3062a6bfbd04755de2/.github/workflows/integration-tests-frontend.yml#L62-L66

Separating out the log improvements from the dependency. We can:

1. Improve the format of Kiali logs gathered by adding more metadata timestamps, name, namespace etc.
2. Log the Kiali logs and/or bookinfo logs per spec file

without adding stern but if we want the ability to continue streaming logs on pod failure then the tool needs to be managed outside of the test suite. Maybe it'd be helpful for someone else to weigh in as well like @jmazzitelli

**Comment by ScriptingShrimp** [comment]
solving https://github.com/kiali/kiali/issues/8492

**Comment by ScriptingShrimp** [comment]
as agreed on a call with @nrfox 
- stern is disabled by default
- stern is controlled by `export CYPRESS_STERN=<true|false>` variable, in the `before` and `after` hooks in cypress (defaults to false)
- if you pass `--stern true` option to the `hack/run-integration-tests.sh`, `download-stern.sh` is now executed as a part of the  script
- for now, output format defaults to JSON, we will extend this into more readable format in future
- edge cases in shell scripts sanitized, failing stern execution will not fail the tests 
- as a first step, we will enable logging for the nightly pipeline only 

requesting review please

File: hack/stern/run-stern.sh [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  run_stern "ratings" "-n bookinfo"
+  run_stern "reviews" "-n bookinfo"
+
+  # check if stern is running, otherwise return non zero exit code, so cypress will notice and 
+  if ! pgrep -f "sternLogs" > /dev/null; then
+    echo "stern is not running. Exiting with error."
+    exit 1
+  fi
+elif [[ "${STOP}" == "true" ]]; then
+    pkill -x stern
```
[the remark this answers is not in the document]
Comment (ScriptingShrimp) [reply hack/stern/run-stern.sh:73]: unfortunately `pkill -x sternLogs` does not work

**Comment by jmazzitelli** [comment]
Regarding the topic of "just get the text logs and you can convert them to JSON later"... how does that work with the new structured logs? I don't see how it can be easy to convert text logs to JSON when the logs don't make it clear which are the ad-hoc messages and which are the structured data. For example, how does one convert this to JSON:

```
2025-07-08T19:44:57Z TRC Adding idle node for workload [kiali] with labels
[map[app:kiali app.kubernetes.io/instance:kiali app.kubernetes.io/name:kiali
app.kubernetes.io/part-of:kiali app.kubernetes.io/version:dev version:dev]]
appender=idleNode group=graph request-id=d1mnaedupccc739f96h0
route=GraphNamespaces route-pattern=/api/namespaces/graph
```

I guess you could look for "X=Y" and assume X is a JSON key and Y is the JSON value and anything else is the message (???) That assumes Y doesn't have separators (like spaces) itself and it assumes actual message string does not also contain X=Y strings - which does happen in some Kiali logs, such as:

```
2025-07-08T19:45:17Z TRC Ignoring validation failure [&{Code:KIA1301 Message:This
workload is not covered by any authorization policy Severity:warning
Path:workload}] for object [/, Kind=workload:oauth2-proxy] in namespace [oauth2-proxy]
```

Converting from plain text to  JSON will not be easy IMO. It certainly would be easiest to just set "logger.log_format" to JSON and be done with it, if our QE frameworks want JSON to begin with.

File: hack/stern/run-stern.sh [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      echo "Unknown argument [$key]. Aborting."
+      exit 1
+      ;;
+  esac
+done
+
+run_stern() {
+  local resource="$1"
+  local namespace="$2"
+  bash -c "exec -a sternLogs \"$SCRIPT_DIR/stern\" $resource $namespace --tail 0 -o json --only-log-lines --max-log-requests 150 >> $LOGFILE 2>&1" &
```
Comment (nrfox) [inline hack/stern/run-stern.sh:49]: I'm still -1 on outputting json logs. It makes the logs harder to read without any benefit right now.
```suggestion
  bash -c "exec -a sternLogs \"$SCRIPT_DIR/stern\" $resource $namespace --tail 0 --only-log-lines --max-log-requests 150 >> $LOGFILE 2>&1" &
```

File: hack/stern/run-stern.sh [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+}
+
+if [[ "${STOP}" == "false" && "${LOGFILE}" != "" ]]; then
+  echo '{"message":"---next scenario in the feature file---"}' >> $LOGFILE
+  # core mesh containers
+  run_stern "kiali" "-n istio-system"
+  run_stern "kiali" "-n kiali-operator"
```
Comment (nrfox) [inline hack/stern/run-stern.sh:56]: How does this behave when the pod doesn't exist? Does it fail? The `kiali-operator` namespace won't exist in CI runs.

File: hack/stern/run-stern.sh [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  run_stern "istiod" "-n istio-system"
+  run_stern "istio-ingressgateway" "-n istio-system"
+
+  # bookinfo application
+  run_stern "details" "-n bookinfo"
+  run_stern "kiali-traffic-generator" "-n bookinfo"
+  run_stern "productpage" "-n bookinfo"
+  run_stern "ratings" "-n bookinfo"
+  run_stern "reviews" "-n bookinfo"
```
Comment (nrfox) [inline hack/stern/run-stern.sh:65]: Looks like you can just specify the namespace to get everything in it: `stern . -n kube-system`

File: frontend/cypress/integration/common/hooks.ts [hunk 6]
```diff
@@ -203,3 +203,24 @@ After({ tags: '@shared-mesh-config' }, () => {
   const patch = '{"spec": {"values": {"pilot": {"env": {"SHARED_MESH_CONFIG": null}}}}}';
   cy.exec(`kubectl patch istio default --type='merge' -p '${patch}'`);
 });
+
+Before(() => {
+  if (Cypress.env('STERN') === true) {
```
[the remark this answers is not in the document]
Comment (ScriptingShrimp) [reply frontend/cypress/integration/common/hooks.ts:208]: Cypress.env returns boolean

File: .github/workflows/integration-tests-frontend.yml [hunk 7]
```diff
@@ -78,3 +82,11 @@ jobs:
       with:
         name: cypress-screenshots-${{ github.job }}
         path: frontend/cypress/screenshots
+
+    - name: Upload cypress logs from stern
+      uses: actions/upload-artifact@v4
+      if: ${{ inputs.stern_logs == true }}
+      with:
+        name: cypress-logs-${{ github.job }}
+        path: hack/stern
```
Comment (nrfox) [inline .github/workflows/integration-tests-frontend.yml:91]: You should only need these on failure. See the other debug info/upload steps. In fact, it'd be nice if this were part of `ci-get-debug-info.sh`.

File: hack/stern/run-stern.sh (hunk 5 above)
Comment (nrfox) [reply hack/stern/run-stern.sh:65]: We can probably even label all the pods we want logs from e.g. `kiali.io/testing: ""` and add a label selector across all namespaces so we only have to run one instance of stern.

**Review by nrfox — CHANGES_REQUESTED** [review]
Mostly just a few suggestions that aren't necessary but I'm still -1 on outputting json logs. They should be just be text otherwise it harms readability.

File: hack/stern/run-stern.sh (hunk 4 above)
Comment (ScriptingShrimp) [reply hack/stern/run-stern.sh:56]:
```
$ ./stern test test --no-follow
$ $?
bash: 0: command not found...
```

Without `--no-follow` process is spawned, but does not output anything and sleeps

File: hack/stern/run-stern.sh (hunk 5 above)
Comment (ScriptingShrimp) [reply hack/stern/run-stern.sh:65]: Noted. Will adjust this in the future, more resources could be added/removed. Now this is bare minimum for upstream and midstream, in private and public pipelines.

File: .github/workflows/integration-tests-frontend.yml (hunk 7 above)
Comment (ScriptingShrimp) [reply .github/workflows/integration-tests-frontend.yml:91]: > You should only need these on failure.

I want to extract "healthy" logs as well. We need good base line to discover anomalies. 

Speaking of `ci-get-debug-info.sh` - runs on the end of the pipeline, we will not get logs from the crashed pods (which were restarted multiple times (as we mentioned earlier).

File: hack/stern/run-stern.sh (hunk 3 above)
Comment (ScriptingShrimp) [reply hack/stern/run-stern.sh:49]: As Mazz mentioned, feeding ETL pipeline with unstructured data is pain. The goal for now is to get good baseline data in machine readable format - so the benefit is we can store and parse the logs in unified format. 

Let's create issue about this flag if you are concerned about readability when using stern logs locally.

**Comment by nrfox** [comment]
> Converting from plain text to JSON will not be easy IMO. It certainly would be easiest to just set "logger.log_format" to JSON and be done with it, if our QE frameworks want JSON to begin with.

We're talking about two different things here. This PR does not change the format that Kiali logs nor does it change any of the other apps (bookinfo, istiod, etc.). `stern -o json` just wraps the text in JSON. It's this:
```
{"message":"2025-07-09T11:41:53Z TRC Found controlplane [istiod/istio-system] on cluster [cluster-default].","nodeName":"ci-worker","namespace":"istio-system","podName":"kiali-6c8599ff6d-g8brp","containerName":"kiali"}
```
vs. this
```
kiali-6c8599ff6d-g8brp kiali 2025-07-09T11:44:54Z TRC Found controlplane [istiod/istio-system] on cluster [cluster-default].
```

> As Mazz mentioned, feeding ETL pipeline with unstructured data is pain. The goal for now is to get good baseline data in machine readable format - so the benefit is we can store and parse the logs in unified format.

This is what is confusing me then. The whole goal of this PR and adding stern is _not_ to make it immediately easier to debug failures in the pipeline but instead to ship _all_ logs somewhere?

File: .github/workflows/integration-tests-frontend.yml (hunk 7 above)
Comment (nrfox) [reply .github/workflows/integration-tests-frontend.yml:91]: There's a limit to how much data can be uploaded to github artifacts. I'm not sure what that limit is but uploading all logs for every run will definitely take up more space. If you want all logs, not to diagnose Kiali failures in CI but for other purposes, you might consider just getting the logs through the GH API.

> Speaking of ci-get-debug-info.sh - runs on the end of the pipeline, we will not get logs from the crashed pods (which were restarted multiple times (as we mentioned earlier).

No I'm saying you can put the stern logs in the same directory that the rest of the debug info is logged to e.g. `--output-directory debug-output`.

File: hack/stern/run-stern.sh (hunk 3 above)
Comment (jmazzitelli) [reply hack/stern/run-stern.sh:49]: Regarding readability of the logs, is it possible to come up with some AI mechanism to help parse and analyse logs? AI is becoming magical - perhaps we can feed logs and ask Claude "tell me what errors are happening and give me suggestions on how to fix it" :) 

 (BTW, this whole "text is more readable than json" is subjective. I, personally, don't think the Kiali text logs are all that much more readable than JSON -- I have a hard time picking out the structured data from the text format logs with my eyes. Is this text message really all that much readable than the json, taking into account the structured part?)

```
2025-07-09T12:16:52Z TRC Build [versionedApp] graph for [5] namespaces [map[default:{default 1m0s false false} dex:{dex 1m0s false false} ingress-nginx:{ingress-nginx 1m0s false false} istio-system:{istio-system 1m0s false true} oauth2-proxy:{oauth2-proxy 1m0s false false}]] group=graph request-id=d1n5rd32f2ls73bvbia0 route=GraphNamespaces route-pattern=/api/namespaces/graph
```

```json
{"level":"trace","route":"GraphNamespaces","route-pattern":"/api/namespaces/graph","group":"graph","request-id":"d1n5s65ntq3c739s4osg","ts":"2025-07-09T12:18:32Z","msg":"Build [versionedApp] graph for [5] namespaces [map[default:{default 1m0s false false} dex:{dex 1m0s false false} ingress-nginx:{ingress-nginx 1m0s false false} istio-system:{istio-system 1m0s false true} oauth2-proxy:{oauth2-proxy 1m0s false false}]]"}
```

**Comment by jmazzitelli** [comment]
I was under the assumption this PR was all about shipping KIALI logs somewhere. If it is about shipping all the logs, like bookinfo and whatever else (istiod logs?), and wrapping THEIR text logs in a json message with a single "msg" key -- well, that seems fairly useless.

**Comment by ScriptingShrimp** [comment]
> Mostly just a few suggestions that aren't necessary but I'm still -1 on outputting json logs. They should be just be text otherwise it harms readability.

hey @nrfox I would prefer to create an issue about this flag. I can provide there some examples how to better read the logs or we can discuss other options for the flag to make the logging more versatile.
